src/main/java/fi/pohina/vinkkilista/data_access/PostgresBookmarkDao.java

```diff
@@ -28,9 +26,9 @@ public PostgresBookmarkDao(ConnectionProvider connProvider) {
      */
     @Override
     public Bookmark findById(String id) {
-        try {
+        try (Connection conn = dataSource.getConnection()) {
```

Comment:
btw for others: this is a [`try-with-resources`](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) statement. It automatically calls `conn.close()` even if an exception is thrown. This way connections are always returned to the connection pool for use.